In [62]:
location_dict = {
    '감우리': 'Gamuri',
    '남제주소내': 'Namjeju Power Plant',
    '무릉리': 'Mureung-ri',
    '부산복합자재창고': 'Busan Composite Material Warehouse',
    '부산수처리장': 'Busan Water Treatment Plant',
    '부산신항': 'Busan New Port',
    '부산역선상주차장': 'Busan Station Overhead Parking Lot',
    '부산운동장': 'Busan Stadium',
    '세화리': 'Sehwa-ri',
    '송당리': 'Songdang-ri',
    '신인천 1_2단계 주차장': 'New Incheon Phase 1_2 Parking Lot',
    '신인천 북측부지': 'New Incheon North Site',
    '신인천 주차장': 'New Incheon Parking Lot',
    '신인천본관주차장': 'New Incheon Main Building Parking Lot',
    '신인천소내': 'New Incheon Power Plant',
    '신인천전망대': 'New Incheon Observatory',
    '신인천해수구취수구': 'New Incheon Seawater Intake',
    '신풍리': 'Shinpung-ri',
    '영월본부': 'Yeongwol Headquarters',
    '영월철도부지': 'Yeongwol Railway Site',
    '와산리': 'Wasan-ri',
    '용수리': 'Yongsu-ri',
    '위미2리': 'Wimi 2-ri',
    '이천D(백사면B)': 'Icheon D (Baeksamyeon B)',
    '이천시 백사면A': 'Icheon-si Baeksamyeon A',
    '익산 다송리': 'Iksan Dasong-ri',
    '인천수산정수장': 'Incheon Susan Water Purification Plant',
    '하동공설운동장': 'Hadong Public Stadium',
    '하동변전소': 'Hadong Substation',
    '하동보건소': 'Hadong Public Health Center',
    '하동정수장': 'Hadong Water Purification Plant',
    '하동하수처리장': 'Hadong Sewage Treatment Plant',
    '행원소수력': 'Haengwon Small Hydropower Plant',
    '화촌주민참여형': 'Hwachon Resident-participatory Plant',
    '부산본부_1': 'Busan HQ_1',
    '부산본부_2': 'Busan HQ_2',
    '삼척소내_1': 'Samcheok Power Plant_1',
    '삼척소내_2': 'Samcheok Power Plant_2',
    '삼척소내_3': 'Samcheok Power Plant_3',
    '삼척소내_4': 'Samcheok Power Plant_4',
    '하동본부_1': 'Hadong HQ_1',
    '하동본부_2': 'Hadong HQ_2',
    '하동본부_3': 'Hadong HQ_3',
    '하동본부_4': 'Hadong HQ_4',
    '하동본부_5': 'Hadong HQ_5',
    '하동본부_6': 'Hadong HQ_6',
}


province_dict = {
    '충청북도': 'Chungcheongbuk-do',
    '제주특별자치도': 'Jeju',
    '부산': 'Busan',
    '인천광역시': 'Incheon',
    '강원도': 'Gangwon-do',
    '경기도': 'Gyeonggi-do',
    '전라북도': 'Jeollabuk-do',
    '경상남도': 'Gyeongsangnam-do'
}


In [36]:
import os
import numpy as np

meta_data = pd.read_csv('/home/local/soominchung_991007/EF/code/tmp/data/solar_energy/meta_data.csv')
result_path = '/home/local/soominchung_991007/EF/code/tmp/result'
folders  = os.listdir(result_path)

def load_prediction_set(result_path, prefix):
    """
    result_path: 파일들이 저장된 폴더 경로
    prefix: 'test', 'valid', 'external_test' 등
    
    리턴: {
        'pred': ndarray,
        'real': ndarray,
        'capacity': ndarray,
        'plant_names': ndarray
    }
    """
    data = {}
    data['pred'] = np.load(os.path.join(result_path, f'{prefix}_predictions_denormalized.npy'),allow_pickle=True)
    data['real'] = np.load(os.path.join(result_path, f'{prefix}_targets_original.npy'),allow_pickle=True)
    data['pred_norm'] = np.load(os.path.join(result_path, f'{prefix}_predictions_normalized.npy'),allow_pickle=True)
    data['real_norm'] = np.load(os.path.join(result_path, f'{prefix}_targets_normalized.npy'),allow_pickle=True)
    data['capacity'] = np.load(os.path.join(result_path, f'{prefix}_capacities.npy'),allow_pickle=True)
    data['plant_names'] = np.load(os.path.join(result_path, f'{prefix}_plant_names.npy'),allow_pickle=True)
    return data


valid_files = {}
test_files = {}
external_files = {}
for ff in folders:
    if ff.split('_')[-2]=='0712' or ff.split('_')[-2]=='0713':
        pth = os.path.join(result_path,ff)
        try:
            valid_files[ff] = load_prediction_set(pth,'valid')
            test_files[ff] = load_prediction_set(pth, 'test')
            external_files[ff] = load_prediction_set(pth, 'external_test')            
        except:
            print(ff)

transformer_256d_model_2n_layers_3e-05lr_16batch_MSELoss_0713_173920
transformer_256d_model_2n_layers_3e-05lr_16batch_MSELoss_0713_102541
transformer_256d_model_2n_layers_3e-05lr_16batch_MSE_PER_Weighted_0713_102717
transformer_256d_model_2n_layers_3e-05lr_16batch_MSE_PER_Weighted_0713_173952
simple_64d_model_1n_layers_0.0001lr_32batch_MSELoss_0712_163830
transformer_256d_model_2n_layers_3e-05lr_16batch_MSELoss_0713_102437
simple_64d_model_1n_layers_0.0001lr_32batch_MSELoss_0712_163739
transformer_256d_model_2n_layers_3e-05lr_16batch_MSELoss_0713_102645
simple_64d_model_1n_layers_0.0001lr_32batch_PredictionErrorRate_0712_165638
transformer_256d_model_2n_layers_3e-05lr_16batch_MSE_PER_Weighted_0713_102821
transformer_256d_model_2n_layers_3e-05lr_16batch_MSELoss_0713_102333
transformer_256d_model_2n_layers_3e-05lr_16batch_MSELoss_0713_102749
transformer_256d_model_2n_layers_3e-05lr_16batch_MSE_PER_Weighted_0713_102509
64d_model_1n_layers_5e-05lr_64batch_MSELoss_0712_130341
transformer_25

In [41]:
import pandas as pd
import numpy as np

def calculate_overall_metrics(pred, real, capacity, plant_names, min_samples=1):
   """
   전체 data에 대한 메트릭 계산
   """
   # Capacity를 예측/실제값과 같은 shape로 확장
   if len(capacity.shape) == 1:
       capacity_expanded = capacity[:, np.newaxis]
   else:
       capacity_expanded = capacity
   
   # 이용률 계산 (10% 이상만 유효)
   utilization_rate = (real / capacity_expanded) * 100
   valid_mask = utilization_rate >= 10.0
   
   # 오차율 계산 (절대 오차 / 용량 * 100)
   error_rate = np.abs(real - pred) / capacity_expanded * 100
   
   # 유효한 데이터만 선택
   masked_error = np.where(valid_mask, error_rate, np.nan)
   
   # 샘플별 평균 메트릭 (시간축 평균)
   sample_metrics = np.nanmean(masked_error, axis=1)
   
   # 유효한 샘플만 선택 (NaN이 아닌 샘플)
   valid_samples = ~np.isnan(sample_metrics)
   
   if np.sum(valid_samples) == 0:
       return np.nan
   
   # 전체 평균 메트릭
   overall_mean_metric = np.mean(sample_metrics[valid_samples])
   
   return overall_mean_metric

# 각 폴더별 메트릭 계산
results = []

for folder_name in valid_files.keys():
   result_dict = {'folder': folder_name}
   
   # valid error
   try:
       valid_err = calculate_overall_metrics(
           valid_files[folder_name]['pred'], 
           valid_files[folder_name]['real'], 
           valid_files[folder_name]['capacity'], 
           valid_files[folder_name]['plant_names']
       )
       result_dict['valid_err'] = valid_err
   except Exception as e:
       print(f"Error processing {folder_name} valid: {e}")
       result_dict['valid_err'] = np.nan
   
   # test error
   try:
       test_err = calculate_overall_metrics(
           test_files[folder_name]['pred'], 
           test_files[folder_name]['real'], 
           test_files[folder_name]['capacity'], 
           test_files[folder_name]['plant_names']
       )
       result_dict['test_err'] = test_err
   except Exception as e:
       print(f"Error processing {folder_name} test: {e}")
       result_dict['test_err'] = np.nan
   
   # external error
   try:
       external_err = calculate_overall_metrics(
           external_files[folder_name]['pred'], 
           external_files[folder_name]['real'], 
           external_files[folder_name]['capacity'], 
           external_files[folder_name]['plant_names']
       )
       result_dict['external_err'] = external_err
   except Exception as e:
       print(f"Error processing {folder_name} external: {e}")
       result_dict['external_err'] = np.nan
   
   results.append(result_dict)

# DataFrame 생성
metrics_df = pd.DataFrame(results)
sorted_valid_result = metrics_df.sort_values('valid_err')#[['folder','valid_err']]
best_value = sorted_valid_result.iloc[0]['folder']
sec_value = sorted_valid_result.iloc[1]['folder']
print('*****best: ',best_value)
print('*****sec: ',sec_value)

# 전체 결과 출력
print("\n=== All Metrics ===")
sorted_valid_result

*****best:  transformer_64d_model_2n_layers_5e-05lr_32batch_PredictionErrorRate_0712_192205
*****sec:  transformer_256d_model_2n_layers_3e-05lr_16batch_MSE_PER_Weighted_0713_121515

=== All Metrics ===


/tmp/ipykernel_1402427/2295398325.py:25: RuntimeWarning: Mean of empty slice
  sample_metrics = np.nanmean(masked_error, axis=1)


folder  valid_err   test_err  \
7   transformer_64d_model_2n_layers_5e-05lr_32batc...   9.119313   8.974371   
29  transformer_256d_model_2n_layers_3e-05lr_16bat...   9.453490   9.369192   
23  transformer_32d_model_2n_layers_5e-05lr_32batc...   9.511523   9.286339   
28  transformer_256d_model_2n_layers_3e-05lr_16bat...   9.511888   9.395543   
21  transformer_256d_model_2n_layers_3e-05lr_16bat...   9.511888   9.395543   
30  transformer_256d_model_2n_layers_3e-05lr_16bat...   9.511888   9.395543   
16  transformer_256d_model_2n_layers_3e-05lr_16bat...   9.539337   9.458037   
11  transformer_256d_model_2n_layers_3e-05lr_16bat...   9.539337   9.458037   
18  transformer_256d_model_2n_layers_3e-05lr_16bat...   9.539337   9.458037   
3   transformer_256d_model_2n_layers_3e-05lr_16bat...   9.551259   9.514921   
15  transformer_64d_model_2n_layers_5e-05lr_32batc...   9.848367   9.697209   
2   transformer_128d_model_2n_layers_5e-05lr_32bat...   9.853698   9.775651   
20  transformer_256d_model_2n_layers_3e-05lr_16bat...  10.146341   9.991990   
4   transformer_256d_model_2n_layers_3e-05lr_16bat...  10.146341   9.991990   
22  transformer_256d_model_2n_layers_3e-05lr_16bat...  10.146341   9.991990   
26  transformer_64d_model_2n_layers_5e-05lr_32batc...  10.329460  10.046054   
9   transformer_64d_model_2n_layers_0.0001lr_32bat...  10.603444  10.520689   
25  transformer_256d_model_2n_layers_3e-05lr_16bat...  10.669928  10.381286   
32  transformer_256d_model_2n_layers_3e-05lr_16bat...  10.669928  10.381286   
31  simple_64d_model_1n_layers_0.0001lr_32batch_Pr...  10.946640  11.445220   
8   transformer_64d_model_2n_layers_3e-05lr_64batc...  11.138549  11.078439   
0   simple_128d_model_1n_layers_0.0001lr_32batch_S...  11.247780  11.835564   
19  simple_256d_model_1n_layers_8e-05lr_16batch_Sm...  11.328520  11.910092   
14  simple_256d_model_1n_layers_8e-05lr_16batch_MS...  11.358224  11.644711   
10  simple_128d_model_1n_layers_0.0001lr_32batch_M...  11.359091  11.672375   
27  simple_64d_model_1n_layers_0.0001lr_32batch_MS...  11.369877  11.679116   
33  simple_64d_model_1n_layers_0.0001lr_32batch_MS...  11.382571  11.726164   
17  simple_32d_model_1n_layers_0.0001lr_32batch_MS...  11.384007  11.653905   
12  simple_64d_model_1n_layers_5e-05lr_64batch_MSE...  11.413203  11.681946   
1   simple_128d_model_1n_layers_0.0001lr_32batch_S...  11.427172  11.859150   
5   simple_64d_model_1n_layers_0.0001lr_32batch_MS...  11.480706  11.751050   
13  simple_64d_model_1n_layers_0.0001lr_32batch_MS...  11.480706  11.751050   
24  simple_64d_model_1n_layers_0.0002lr_32batch_MS...  11.512854  11.676981   
6   simple_128d_model_1n_layers_0.0001lr_32batch_M...  11.522573  11.778245   

    external_err  
7      12.926437  
29     14.534435  
23     14.785045  
28     14.549829  
21     14.549829  
30     14.549829  
16     14.391454  
11     14.391454  
18     14.391454  
3      14.550960  
15     15.325718  
2      14.359612  
20     14.703079  
4      14.703079  
22     14.703079  
26     15.531795  
9      15.410869  
25     15.182146  
32     15.182146  
31     14.590693  
8      16.758842  
0      15.896358  
19     16.129358  
14     15.788114  
10     15.332946  
27     15.616500  
33     15.892604  
17     15.862585  
12     15.764869  
1      16.494242  
5      15.772916  
13     15.772916  
24     15.641489  
6      15.606658

In [46]:
print(meta_data['name'].unique())

['감우리' '남제주소내' '무릉리' '부산복합자재창고' '부산수처리장' '부산신항' '부산역선상주차장' '부산운동장' '세화리'
 '송당리' '신인천 1_2단계 주차장' '신인천 북측부지' '신인천 주차장' '신인천본관주차장' '신인천소내' '신인천전망대'
 '신인천해수구취수구' '신풍리' '영월본부' '영월철도부지' '와산리' '용수리' '위미2리' '이천D(백사면B)'
 '이천시 백사면A' '익산 다송리' '인천수산정수장' '하동공설운동장' '하동변전소' '하동보건소' '하동정수장' '하동하수처리장'
 '행원소수력' '화촌주민참여형' '부산본부_1' '부산본부_2' '삼척소내_1' '삼척소내_2' '삼척소내_3' '삼척소내_4'
 '하동본부_1' '하동본부_2' '하동본부_3' '하동본부_4' '하동본부_5' '하동본부_6']


### 결과 분석

In [ ]:
import matplotlib.pyplot as plt
import numpy as np
import pandas as pd
from datetime import datetime

# best_value 폴더의 데이터 로드
best_data = {
    'valid': valid_files[best_value],
    'test': test_files[best_value], 
    'external': external_files[best_value]
}

In [51]:
# =================================================================
# 1. 오류 분석을 위한 샘플 선택 및 플롯
# =================================================================
import os

def get_sample_metrics_with_indices(pred, real, capacity, plant_names):
   """샘플별 메트릭과 인덱스 반환"""
   capacity_expanded = capacity[:, np.newaxis]
   utilization_rate = (real / capacity_expanded) * 100
   valid_mask = utilization_rate >= 10.0
   error_rate = np.abs(real - pred) / capacity_expanded * 100
   masked_error = np.where(valid_mask, error_rate, np.nan)
   sample_metrics = np.nanmean(masked_error, axis=1)
   valid_samples = ~np.isnan(sample_metrics)
   
   return sample_metrics, valid_samples

def plot_sample_analysis(data_type='valid'):
   """특정 데이터 타입에 대한 샘플 분석 플롯"""
   data = best_data[data_type]
   sample_metrics, valid_samples = get_sample_metrics_with_indices(
       data['pred'], data['real'], data['capacity'], data['plant_names']
   )
   
   valid_metrics = sample_metrics[valid_samples]
   valid_indices = np.where(valid_samples)[0]
   
   # 인덱스 선택
   sorted_indices = np.argsort(valid_metrics)
   
   # 최고 오류 5개
   worst_5_idx = valid_indices[sorted_indices[-5:]]
   # 최저 오류 5개  
   best_5_idx = valid_indices[sorted_indices[:5]]
   # 랜덤 5개
   np.random.seed(42)
   random_5_idx = np.random.choice(valid_indices, 5, replace=False)
   
   # 저장 경로 생성
   save_dir = f'/home/local/soominchung_991007/EF/code/tmp/result_analysis/best/{data_type}'
   os.makedirs(save_dir, exist_ok=True)
   
   # 플롯 생성
   categories = [('Best_5_Lowest_Error', best_5_idx), 
                 ('Worst_5_Highest_Error', worst_5_idx),
                 ('Random_5', random_5_idx)]
   
   for cat_name, indices in categories:
       fig, axs = plt.subplots(5, 1, figsize=(12, 15))
       fig.suptitle(f'{data_type.upper()} - {cat_name.replace("_", " ")}', fontsize=16)
       
       for i, idx in enumerate(indices):
           hours = np.arange(24)
           real_vals = data['real'][idx]
           pred_vals = data['pred'][idx]
           capacity = data['capacity'][idx]
           plant_name = location_dict[data['plant_names'][idx]]
           error = sample_metrics[idx]
           
           # 이용률 계산 및 유효성 마스크
           utilization_rate = (real_vals / capacity) * 100
           valid_mask = utilization_rate >= 10.0
           
           # 10% 이상인 점들 (사용되는 점들)
           axs[i].plot(hours[valid_mask], real_vals[valid_mask], 'o-', alpha=0.7, 
                      label='Actual (Used)', color='C0', markersize=4)
           axs[i].plot(hours[valid_mask], pred_vals[valid_mask], 's--', alpha=0.7, 
                      label='Predicted (Used)', color='C1', markersize=4)
           
           # 10% 미만인 점들 (사용되지 않는 점들)
           if np.sum(~valid_mask) > 0:
               axs[i].plot(hours[~valid_mask], real_vals[~valid_mask], 'o', alpha=0.3, 
                          label='Actual (Unused)', color='gray', markersize=3)
               axs[i].plot(hours[~valid_mask], pred_vals[~valid_mask], 's', alpha=0.3, 
                          label='Predicted (Unused)', color='lightgray', markersize=3)
           
           # 10% 이용률 기준선 (빨간 가로선)
           threshold_power = capacity * 0.1
           axs[i].axhline(y=threshold_power, color='red', linestyle='-', alpha=0.7, 
                         linewidth=2, label=f'10% Threshold ({threshold_power:.1f} MW)')
           
           axs[i].set_title(f'{plant_name} ({capacity:.1f} MW) - Error: {error:.2f}%')
           axs[i].set_xlabel('Hour')
           axs[i].set_ylabel('Power (MW)')
           axs[i].set_xticks([0, 6, 12, 18, 23])
           axs[i].legend(fontsize=8)
           axs[i].grid(True, alpha=0.3)
       
       plt.tight_layout()
       
       # 파일 저장
       save_path = os.path.join(save_dir, f'{cat_name}.png')
       plt.savefig(save_path, dpi=300, bbox_inches='tight')
       plt.close()  # 메모리 절약을 위해 닫기
       print(f"Saved: {save_path}")

# 각 데이터 타입별 플롯 실행
for data_type in ['valid', 'test', 'external']:
   plot_sample_analysis(data_type)

/tmp/ipykernel_1402427/1481092424.py:13: RuntimeWarning: Mean of empty slice
  sample_metrics = np.nanmean(masked_error, axis=1)


Saved: /home/local/soominchung_991007/EF/code/tmp/result_analysis/best/valid/Best_5_Lowest_Error.png
Saved: /home/local/soominchung_991007/EF/code/tmp/result_analysis/best/valid/Worst_5_Highest_Error.png
Saved: /home/local/soominchung_991007/EF/code/tmp/result_analysis/best/valid/Random_5.png
Saved: /home/local/soominchung_991007/EF/code/tmp/result_analysis/best/test/Best_5_Lowest_Error.png
Saved: /home/local/soominchung_991007/EF/code/tmp/result_analysis/best/test/Worst_5_Highest_Error.png
Saved: /home/local/soominchung_991007/EF/code/tmp/result_analysis/best/test/Random_5.png
Saved: /home/local/soominchung_991007/EF/code/tmp/result_analysis/best/external/Best_5_Lowest_Error.png
Saved: /home/local/soominchung_991007/EF/code/tmp/result_analysis/best/external/Worst_5_Highest_Error.png
Saved: /home/local/soominchung_991007/EF/code/tmp/result_analysis/best/external/Random_5.png


In [72]:
# =================================================================
# 2. 발전소별 메타데이터 매핑 및 지역별 성능 분석
# =================================================================

def analyze_performance_by_region():
    """발전소별 메타데이터를 매핑하고 지역별 성능 분석"""
    
    # 메타데이터 매핑 준비
    meta_dict = {}
    if 'meta_data' in globals() and meta_data is not None:
        for _, row in meta_data.iterrows():
            meta_dict[row['name']] = {
                'location_name': row.get('location_name', 'Unknown'),
                '광역시': row.get('광역시', 'Unknown'), 
                '준공년도': row.get('준공년도', 'Unknown')
            }
    else:
        print("meta_data not available. Using plant names as is.")
        return None
    
    regional_results = []
    
    for data_type in ['valid', 'test', 'external']:
        data = best_data[data_type]
        
        # 각 발전소별 성능 계산
        capacity_expanded = data['capacity'][:, np.newaxis]
        utilization_rate = (data['real'] / capacity_expanded) * 100
        valid_mask = utilization_rate >= 10.0
        error_rate = np.abs(data['real'] - data['pred']) / capacity_expanded * 100
        masked_error = np.where(valid_mask, error_rate, np.nan)
        sample_metrics = np.nanmean(masked_error, axis=1)
        
        # 유효한 샘플만 선택
        valid_samples = ~np.isnan(sample_metrics)
        
        for i, plant_name in enumerate(data['plant_names']):
            if not valid_samples[i]:
                continue
                
            # 메타데이터 매핑
            plant_meta = meta_dict.get(plant_name, {
                'location_name': plant_name,
                '광역시': 'Unknown',
                '준공년도': 'Unknown'
            })
            
            regional_results.append({
                'data_type': data_type,
                'plant_name': plant_name,
                'location_name': plant_meta['location_name'],
                '광역시': plant_meta['광역시'],
                '준공년도': plant_meta['준공년도'],
                'capacity': data['capacity'][i],
                'error_rate': sample_metrics[i],
                'sample_index': i
            })
    
    # DataFrame 생성
    plant_results_df = pd.DataFrame(regional_results)
    
    # 지역별(광역시별) 성능 집계
    regional_summary = []
    
    for data_type in ['valid', 'test', 'external']:
        type_data = plant_results_df[plant_results_df['data_type'] == data_type]
        
        regional_stats = type_data.groupby('광역시').agg({
            'error_rate': ['mean', 'std'],
            'capacity': ['mean', 'sum'],
            'plant_name': 'nunique'  # 고유한 발전소 개수
        }).round(3)
        
        # 컬럼명 정리
        regional_stats.columns = ['mean_error', 'std_error', 'mean_capacity', 'total_capacity', 'plant_count']
        regional_stats['data_type'] = data_type
        regional_stats['region'] = regional_stats.index
        regional_stats = regional_stats.reset_index(drop=True)
        
        regional_summary.append(regional_stats)
    
    regional_summary_df = pd.concat(regional_summary, ignore_index=True)
    
    print("=== 발전소별 상세 결과 (처음 10개) ===")
    print(plant_results_df.head(10).to_string(index=False))
    
    print("\n=== 지역별(광역시별) 성능 요약 ===")
    print(regional_summary_df.to_string(index=False))
    
    # 지역별 성능 시각화
    fig, axes = plt.subplots(1, 3, figsize=(18, 6))
    
    for i, data_type in enumerate(['valid', 'test', 'external']):
        type_summary = regional_summary_df[regional_summary_df['data_type'] == data_type]
        type_summary = type_summary.sort_values('mean_error')
        
        bars = axes[i].bar(range(len(type_summary)), type_summary['mean_error'], 
                          yerr=type_summary['std_error'], capsize=5)
        
        axes[i].set_title(f'{data_type.upper()} - Regional Performance')
        axes[i].set_ylabel('Mean Error Rate (%)')
        axes[i].set_xlabel('Region')
        axes[i].set_xticks(range(len(type_summary)))
        region_list = []
        region_series = type_summary['region']
        for region in region_series:
            english_name = province_dict.get(region, 'Unknown')  # 매핑되지 않으면 'Unknown'
            region_list.append(english_name)
        axes[i].set_xticklabels(region_list, rotation=45, ha='right')
        #axes[i].set_xticklabels(type_summary['region'], rotation=45, ha='right')
        axes[i].grid(True, alpha=0.3)
        
        # 발전소 개수 표시
        for j, (bar, count) in enumerate(zip(bars, type_summary['plant_count'])):
            axes[i].text(bar.get_x() + bar.get_width()/2, bar.get_height() + 0.1,
                        f'n={count}', ha='center', va='bottom', fontsize=8)
    
    plt.tight_layout()
    
    # 저장
    save_dir = '/home/local/soominchung_991007/EF/code/tmp/result_analysis/best'
    os.makedirs(save_dir, exist_ok=True)
    plt.savefig(os.path.join(save_dir, 'regional_performance.png'), dpi=300, bbox_inches='tight')
    plt.close()
    
    # CSV 저장
    plant_results_df.to_csv(os.path.join(save_dir, 'plant_level_results.csv'), index=False)
    regional_summary_df.to_csv(os.path.join(save_dir, 'regional_summary.csv'), index=False)
    
    print(f"\n결과 저장 완료: {save_dir}")
    
    return plant_results_df, regional_summary_df

# 분석 실행
if 'meta_data' in globals():
    plant_results, regional_summary = analyze_performance_by_region()
else:
    print("meta_data가 로드되지 않았습니다. meta_data를 먼저 로드해주세요.")

/tmp/ipykernel_1402427/444543340.py:32: RuntimeWarning: Mean of empty slice
  sample_metrics = np.nanmean(masked_error, axis=1)


=== 발전소별 상세 결과 (처음 10개) ===
data_type plant_name location_name   광역시  준공년도  capacity  error_rate  sample_index
    valid    신인천 주차장            인천 인천광역시  2019     603.0    5.463086             0
    valid    신인천 주차장            인천 인천광역시  2019     603.0   10.196657             1
    valid    신인천 주차장            인천 인천광역시  2019     603.0   12.935080             2
    valid    신인천 주차장            인천 인천광역시  2019     603.0    3.760879             3
    valid    신인천 주차장            인천 인천광역시  2019     603.0   27.599075             4
    valid    신인천 주차장            인천 인천광역시  2019     603.0    5.367375             5
    valid    신인천 주차장            인천 인천광역시  2019     603.0    9.186459             6
    valid    신인천 주차장            인천 인천광역시  2019     603.0    4.618058             7
    valid    신인천 주차장            인천 인천광역시  2019     603.0    7.112002             8
    valid    신인천 주차장            인천 인천광역시  2019     603.0   13.184205             9

=== 지역별(광역시별) 성능 요약 ===
 mean_error  std_error  mean_capac

In [108]:
# =================================================================
# 지역별 성능 차이 통계 분석
# =================================================================

def analyze_regional_performance_statistics(regional_summary_df, plant_results_df):
    """지역별 성능 차이 통계 분석"""
    import scipy.stats as stats
    from scipy.stats import f_oneway, kruskal
    from itertools import combinations
    
    print("=== 지역별 성능 차이 통계 분석 ===")
    
    # 각 data_type별로 분석
    for data_type in ['valid', 'test', 'external']:
        print(f"\n{'='*50}")
        print(f"{data_type.upper()} 데이터 지역별 분석")
        print(f"{'='*50}")
        
        # 해당 data_type의 개별 발전소 데이터 추출
        type_data = plant_results_df[plant_results_df['data_type'] == data_type].copy()
        
        if len(type_data) == 0:
            print(f"No data for {data_type}")
            continue
        
        # 지역별 데이터 분리 (Unknown 제외)
        regions = type_data['광역시'].unique()
        regions = [r for r in regions if r != 'Unknown']
        
        if len(regions) < 2:
            print(f"Not enough regions for statistical analysis (only {len(regions)} regions)")
            continue
        
        print(f"분석 대상 지역: {regions}")
        
        # 지역별 error_rate 데이터 수집
        regional_data = {}
        for region in regions:
            region_errors = type_data[type_data['광역시'] == region]['error_rate'].values
            if len(region_errors) > 0:
                regional_data[region] = region_errors
                print(f"  {region}: n={len(region_errors)}, mean={np.mean(region_errors):.3f}%, std={np.std(region_errors):.3f}%")
        
        # 통계 검정을 위한 데이터 준비
        region_groups = list(regional_data.values())
        region_names = list(regional_data.keys())
        
        if len(region_groups) >= 2:
            # 1. ANOVA 검정
            try:
                f_stat, p_value_anova = f_oneway(*region_groups)
                print(f"\n📊 ANOVA 검정:")
                print(f"  F-statistic: {f_stat:.4f}")
                print(f"  p-value: {p_value_anova:.6f}")
                print(f"  결과: {'지역별 차이 있음' if p_value_anova < 0.05 else '지역별 차이 없음'} (α=0.05)")
            except Exception as e:
                print(f"\n❌ ANOVA 검정 실패: {e}")
            
            # 2. Kruskal-Wallis 검정 (비모수)
            try:
                kw_stat, p_value_kw = kruskal(*region_groups)
                print(f"\n📊 Kruskal-Wallis 검정:")
                print(f"  H-statistic: {kw_stat:.4f}")
                print(f"  p-value: {p_value_kw:.6f}")
                print(f"  결과: {'지역별 차이 있음' if p_value_kw < 0.05 else '지역별 차이 없음'} (α=0.05)")
            except Exception as e:
                print(f"\n❌ Kruskal-Wallis 검정 실패: {e}")
            
            # 3. 사후 검정 (Pairwise comparison) - 유의한 경우만
            if len(region_groups) >= 2 and (p_value_anova < 0.05 if 'p_value_anova' in locals() else True):
                print(f"\n📊 지역간 쌍별 비교 (Mann-Whitney U 검정):")
                
                pairwise_results = []
                for i, j in combinations(range(len(region_names)), 2):
                    region1, region2 = region_names[i], region_names[j]
                    data1, data2 = region_groups[i], region_groups[j]
                    
                    try:
                        statistic, p_value = stats.mannwhitneyu(data1, data2, alternative='two-sided')
                        
                        # Bonferroni 보정
                        n_comparisons = len(list(combinations(range(len(region_names)), 2)))
                        p_adjusted = min(p_value * n_comparisons, 1.0)
                        
                        pairwise_results.append({
                            'region1': region1,
                            'region2': region2,
                            'mean1': np.mean(data1),
                            'mean2': np.mean(data2),
                            'p_value': p_value,
                            'p_adjusted': p_adjusted,
                            'significant': p_adjusted < 0.05
                        })
                        
                        print(f"  {region1} vs {region2}:")
                        print(f"    평균: {np.mean(data1):.3f}% vs {np.mean(data2):.3f}%")
                        print(f"    p-value: {p_value:.6f} (adjusted: {p_adjusted:.6f})")
                        print(f"    결과: {'유의한 차이' if p_adjusted < 0.05 else '차이 없음'}")
                        print()
                        
                    except Exception as e:
                        print(f"  {region1} vs {region2}: 검정 실패 ({e})")
                
                # 가장 유의한 차이들 요약
                significant_pairs = [r for r in pairwise_results if r['significant']]
                if significant_pairs:
                    print(f"🔍 유의한 지역별 차이 ({len(significant_pairs)}개 쌍):")
                    significant_pairs.sort(key=lambda x: x['p_adjusted'])
                    for pair in significant_pairs[:5]:  # 상위 5개만 표시
                        diff = abs(pair['mean1'] - pair['mean2'])
                        print(f"  {pair['region1']} vs {pair['region2']}: {diff:.3f}%p 차이 (p={pair['p_adjusted']:.4f})")
                else:
                    print("🔍 Bonferroni 보정 후 유의한 지역별 차이 없음")
            
            # 4. 효과 크기 (Cohen's f)
            try:
                # 전체 평균
                all_data = np.concatenate(region_groups)
                grand_mean = np.mean(all_data)
                
                # 집단 간 분산
                group_means = [np.mean(group) for group in region_groups]
                group_sizes = [len(group) for group in region_groups]
                
                between_var = sum(n * (mean - grand_mean)**2 for n, mean in zip(group_sizes, group_means)) / sum(group_sizes)
                within_var = np.var(all_data, ddof=1)
                
                cohens_f = np.sqrt(between_var / within_var)
                
                effect_size = "작음" if cohens_f < 0.25 else "중간" if cohens_f < 0.40 else "큰"
                
                print(f"\n📊 효과 크기:")
                print(f"  Cohen's f: {cohens_f:.4f} ({effect_size} 효과)")
                
            except Exception as e:
                print(f"\n❌ 효과 크기 계산 실패: {e}")
    
    # 시각화
    plt.figure(figsize=(20, 12))
    
    for i, data_type in enumerate(['valid', 'test', 'external']):
        type_data = plant_results_df[plant_results_df['data_type'] == data_type].copy()
        
        if len(type_data) == 0:
            continue
        
        # 지역별 박스플롯
        plt.subplot(2, 3, i+1)
        
        regions = [r for r in type_data['광역시'].unique() if r != 'Unknown']
        r_name =[]
        for r in regions:
            r_name.append(province_dict[r])        
        region_data = [type_data[type_data['광역시'] == region]['error_rate'].values for region in regions]
        
        if len(regions) > 1:
            bp = plt.boxplot(region_data, labels=r_name, patch_artist=True)
            
            # 색상 설정
            colors = plt.cm.Set3(np.linspace(0, 1, len(regions)))
            for patch, color in zip(bp['boxes'], colors):
                patch.set_facecolor(color)
                patch.set_alpha(0.7)
        
        plt.title(f'{data_type.upper()} - Regional Error Distribution', fontsize=14, fontweight='bold')
        plt.ylabel('Error Rate (%)', fontsize=12)
        plt.xlabel('Region', fontsize=12)
        plt.xticks(rotation=45, ha='right')
        plt.grid(True, alpha=0.3)
        
        # 바이올린 플롯
        plt.subplot(2, 3, i+4)
        
        if len(regions) > 1:
            parts = plt.violinplot(region_data, positions=range(1, len(regions)+1), showmeans=True)
            
            # 색상 설정
            for pc, color in zip(parts['bodies'], colors):
                pc.set_facecolor(color)
                pc.set_alpha(0.7)
        
        plt.title(f'{data_type.upper()} - Regional Error Density', fontsize=14, fontweight='bold')
        plt.ylabel('Error Rate (%)', fontsize=12)
        plt.xlabel('Region', fontsize=12)

        plt.xticks(range(1, len(regions)+1), r_name, rotation=45, ha='right')
        plt.grid(True, alpha=0.3)
    
    plt.tight_layout()
    
    # 저장
    save_dir = '/home/local/soominchung_991007/EF/code/tmp/result_analysis/best'
    plt.savefig(os.path.join(save_dir, 'regional_statistical_analysis.png'), dpi=300, bbox_inches='tight')
    plt.close()
    
    print(f"\n📁 결과 저장 완료: {save_dir}/regional_statistical_analysis.png")

# 실행
if 'regional_summary' in globals() and 'plant_results' in globals():
    analyze_regional_performance_statistics(regional_summary, plant_results)
else:
    print("regional_summary 또는 plant_results가 없습니다.")

=== 지역별 성능 차이 통계 분석 ===

VALID 데이터 지역별 분석
분석 대상 지역: ['인천광역시', '경기도', '충청북도', '경상남도', '강원도', '부산']
  인천광역시: n=386, mean=9.595%, std=5.854%
  경기도: n=324, mean=10.075%, std=5.179%
  충청북도: n=150, mean=10.386%, std=5.940%
  경상남도: n=1752, mean=10.746%, std=6.031%
  강원도: n=1568, mean=9.171%, std=6.349%
  부산: n=2304, mean=7.550%, std=5.275%

📊 ANOVA 검정:
  F-statistic: 64.9129
  p-value: 0.000000
  결과: 지역별 차이 있음 (α=0.05)

📊 Kruskal-Wallis 검정:
  H-statistic: 428.3825
  p-value: 0.000000
  결과: 지역별 차이 있음 (α=0.05)

📊 지역간 쌍별 비교 (Mann-Whitney U 검정):
  인천광역시 vs 경기도:
    평균: 9.595% vs 10.075%
    p-value: 0.017178 (adjusted: 0.257675)
    결과: 차이 없음

  인천광역시 vs 충청북도:
    평균: 9.595% vs 10.386%
    p-value: 0.031275 (adjusted: 0.469130)
    결과: 차이 없음

  인천광역시 vs 경상남도:
    평균: 9.595% vs 10.746%
    p-value: 0.000042 (adjusted: 0.000623)
    결과: 유의한 차이

  인천광역시 vs 강원도:
    평균: 9.595% vs 9.171%
    p-value: 0.039691 (adjusted: 0.595369)
    결과: 차이 없음

  인천광역시 vs 부산:
    평균: 9.595% vs 7.550%
    p-value: 0.00

/tmp/ipykernel_1402427/1026000378.py:157: MatplotlibDeprecationWarning: The 'labels' parameter of boxplot() has been renamed 'tick_labels' since Matplotlib 3.9; support for the old name will be dropped in 3.11.
  bp = plt.boxplot(region_data, labels=r_name, patch_artist=True)
/tmp/ipykernel_1402427/1026000378.py:157: MatplotlibDeprecationWarning: The 'labels' parameter of boxplot() has been renamed 'tick_labels' since Matplotlib 3.9; support for the old name will be dropped in 3.11.
  bp = plt.boxplot(region_data, labels=r_name, patch_artist=True)
/tmp/ipykernel_1402427/1026000378.py:157: MatplotlibDeprecationWarning: The 'labels' parameter of boxplot() has been renamed 'tick_labels' since Matplotlib 3.9; support for the old name will be dropped in 3.11.
  bp = plt.boxplot(region_data, labels=r_name, patch_artist=True)



📁 결과 저장 완료: /home/local/soominchung_991007/EF/code/tmp/result_analysis/best/regional_statistical_analysis.png


In [98]:
# =================================================================
# plant_results를 data_type별로 발전소별 평균 error_rate 파일로 저장
# =================================================================

def save_plant_average_by_datatype(plant_results_df):
    """data_type별로 발전소별 평균 error_rate 파일 저장"""
    
    save_dir = '/home/local/soominchung_991007/EF/code/tmp/result_analysis/best'
    os.makedirs(save_dir, exist_ok=True)
    
    for data_type in ['valid', 'test', 'external']:
        # 해당 data_type 데이터 필터링
        type_data = plant_results_df[plant_results_df['data_type'] == data_type].copy()
        
        if len(type_data) == 0:
            print(f"No data found for {data_type}")
            continue
        
        # 발전소별 평균 error_rate 계산
        plant_avg = type_data.groupby('plant_name').agg({
            'location_name': 'first',  # 첫 번째 값 사용 (동일해야 함)
            '광역시': 'first',
            '준공년도': 'first', 
            'capacity': 'first',
            'error_rate': ['mean', 'std', 'count']  # 평균, 표준편차, 샘플 수
        }).round(3)
        
        # 컬럼명 정리
        plant_avg.columns = ['location_name', '광역시', '준공년도', 'capacity', 
                            'mean_error_rate', 'std_error_rate', 'sample_count']
        
        # plant_name을 인덱스에서 컬럼으로 이동
        plant_avg = plant_avg.reset_index()
        
        # 정렬 (error_rate 낮은 순)
        plant_avg = plant_avg.sort_values('mean_error_rate')
        
        # 파일 저장
        filename = f'{data_type}_plant_average_performance.csv'
        filepath = os.path.join(save_dir, filename)
        plant_avg.to_csv(filepath, index=False, encoding='utf-8-sig')
        
        print(f"Saved {len(plant_avg)} plants for {data_type}: {filepath}")
        print(f"  Best plant: {plant_avg.iloc[0]['plant_name']} ({plant_avg.iloc[0]['mean_error_rate']:.3f}%)")
        print(f"  Worst plant: {plant_avg.iloc[-1]['plant_name']} ({plant_avg.iloc[-1]['mean_error_rate']:.3f}%)")
        print()

# 실행
if 'plant_results' in globals():
    save_plant_average_by_datatype(plant_results)
else:
    print("plant_results가 없습니다.")

Saved 30 plants for valid: /home/local/soominchung_991007/EF/code/tmp/result_analysis/best/valid_plant_average_performance.csv
  Best plant: 부산본부_1 (4.097%)
  Worst plant: 하동정수장 (15.558%)

Saved 30 plants for test: /home/local/soominchung_991007/EF/code/tmp/result_analysis/best/test_plant_average_performance.csv
  Best plant: 부산본부_1 (3.955%)
  Worst plant: 하동정수장 (18.037%)

Saved 7 plants for external: /home/local/soominchung_991007/EF/code/tmp/result_analysis/best/external_plant_average_performance.csv
  Best plant: 신인천 북측부지 (9.373%)
  Worst plant: 신인천전망대 (20.613%)



In [114]:
a = pd.read_csv('/home/local/soominchung_991007/EF/code/tmp/result_analysis/best/external_plant_average_performance.csv')
a

plant_name location_name      광역시  준공년도  capacity  mean_error_rate  \
0       신인천 북측부지            인천    인천광역시  2020     365.4            9.373   
1  신인천 1_2단계 주차장            인천    인천광역시  2020     296.1            9.507   
2           부산신항            부산       부산  2011     115.0           11.126   
3       부산역선상주차장            부산       부산  2020     834.4           11.131   
4           위미2리           서귀포  제주특별자치도  2018     995.0           12.234   
5            무릉리           서귀포  제주특별자치도  2021     809.0           13.459   
6         신인천전망대            인천    인천광역시  2017    1742.0           20.613   

   std_error_rate  sample_count  
0           5.346           408  
1           5.929           400  
2           6.308           550  
3           6.582           574  
4           7.410           576  
5           7.612           626  
6          11.332           638

In [109]:
a = pd.read_csv('/home/local/soominchung_991007/EF/code/tmp/result_analysis/best/test_plant_average_performance.csv')
a

plant_name location_name    광역시  준공년도  capacity  mean_error_rate  \
0      부산본부_1            부산     부산  2008     390.0            3.955   
1        영월본부            영월    강원도  2009      76.0            4.098   
2      하동본부_2            진주   경상남도  2010    1901.0            6.930   
3     하동공설운동장            진주   경상남도  2011     370.0            7.367   
4     하동하수처리장            진주   경상남도  2011      55.0            8.533   
5    부산복합자재창고            부산     부산  2012     187.0            8.589   
6      하동본부_3            진주   경상남도  2010     601.0            8.706   
7      영월철도부지            영월    강원도  2018    1191.0            8.790   
8       부산운동장            부산     부산  2017    1400.0            8.807   
9     화촌주민참여형            진주   경상남도  2021     999.0            8.868   
10     부산수처리장            부산     부산  2017     111.0            9.081   
11      신인천소내            인천  인천광역시  2012     200.0            9.089   
12     부산본부_2            부산     부산  2018     502.0            9.178   
13     하동본부_1            진주   경상남도  2008     998.0            9.313   
14  이천D(백사면B)            이천    경기도  2021     999.0            9.642   
15    신인천 주차장            인천  인천광역시  2019     603.0            9.699   
16    인천수산정수장            인천  인천광역시  2011     998.0            9.817   
17  신인천해수구취수구            인천  인천광역시  2018     907.0           10.099   
18     삼척소내_1            동해    강원도  2017     999.0           10.188   
19     삼척소내_4            동해    강원도  2019    2599.0           10.208   
20     하동본부_6            진주   경상남도  2020     998.0           10.310   
21      하동보건소            진주   경상남도  2011      40.0           10.316   
22     삼척소내_2            동해    강원도  2018     990.0           10.339   
23     하동본부_4            진주   경상남도  2017     998.0           10.564   
24     하동본부_5            진주   경상남도  2018     998.0           10.602   
25   이천시 백사면A            이천    경기도  2021    1834.0           10.903   
26      하동변전소            진주   경상남도  2011      48.0           11.050   
27     삼척소내_3            동해    강원도  2018    2002.0           11.216   
28        감우리            충주   충청북도  2021     554.0           11.455   
29      하동정수장            진주   경상남도  2011      93.0           18.037   

    std_error_rate  sample_count  
0            2.540           610  
1            2.262           250  
2            4.433           146  
3            4.279           150  
4            4.832           154  
5            5.730           696  
6            6.053           148  
7            5.162           268  
8            6.247           374  
9            6.360           156  
10           5.782           218  
11           4.796            98  
12           5.461           252  
13           5.855           152  
14           5.454           196  
15           5.043            98  
16           5.635           100  
17           4.752           100  
18           7.767           258  
19           8.270           256  
20           6.652           154  
21           5.306           154  
22           7.936           258  
23           6.956           154  
24           7.491           154  
25           7.654           142  
26           5.510           154  
27           8.107           254  
28           9.024           138  
29           7.758           154

In [115]:
a['sample_count']

0    408
1    400
2    550
3    574
4    576
5    626
6    638
Name: sample_count, dtype: int64

In [118]:
import numpy as np
a = np.load('/home/local/soominchung_991007/EF/code/tmp/data/normalized_forecast_modeling_data/14시/external_test/X_meta.npy')
a.shape


(2017, 10)

In [99]:
a = pd.read_csv('/home/local/soominchung_991007/EF/code/tmp/result_analysis/best/external_plant_average_performance.csv')
a

plant_name location_name      광역시  준공년도  capacity  mean_error_rate  \
0       신인천 북측부지            인천    인천광역시  2020     365.4            9.373   
1  신인천 1_2단계 주차장            인천    인천광역시  2020     296.1            9.507   
2           부산신항            부산       부산  2011     115.0           11.126   
3       부산역선상주차장            부산       부산  2020     834.4           11.131   
4           위미2리           서귀포  제주특별자치도  2018     995.0           12.234   
5            무릉리           서귀포  제주특별자치도  2021     809.0           13.459   
6         신인천전망대            인천    인천광역시  2017    1742.0           20.613   

   std_error_rate  sample_count  
0           5.346           408  
1           5.929           400  
2           6.308           550  
3           6.582           574  
4           7.410           576  
5           7.612           626  
6          11.332           638

In [111]:
a = pd.read_csv('/home/local/soominchung_991007/EF/code/tmp/result_analysis/best/valid_plant_average_performance.csv')
a

plant_name location_name    광역시  준공년도  capacity  mean_error_rate  \
0      부산본부_1            부산     부산  2008     390.0            4.097   
1        영월본부            영월    강원도  2009      76.0            4.734   
2      부산수처리장            부산     부산  2017     111.0            7.793   
3     하동공설운동장            진주   경상남도  2011     370.0            8.437   
4      하동본부_2            진주   경상남도  2010    1901.0            8.529   
5    부산복합자재창고            부산     부산  2012     187.0            8.905   
6     하동하수처리장            진주   경상남도  2011      55.0            8.981   
7      삼척소내_1            동해    강원도  2017     999.0            9.096   
8   신인천해수구취수구            인천  인천광역시  2018     907.0            9.115   
9       부산운동장            부산     부산  2017    1400.0            9.138   
10     부산본부_2            부산     부산  2018     502.0            9.320   
11    신인천 주차장            인천  인천광역시  2019     603.0            9.606   
12     하동본부_1            진주   경상남도  2008     998.0            9.650   
13    인천수산정수장            인천  인천광역시  2011     998.0            9.693   
14     삼척소내_4            동해    강원도  2019    2599.0            9.812   
15     삼척소내_3            동해    강원도  2018    2002.0            9.893   
16  이천D(백사면B)            이천    경기도  2021     999.0            9.950   
17      신인천소내            인천  인천광역시  2012     200.0            9.983   
18    화촌주민참여형            진주   경상남도  2021     999.0           10.214   
19   이천시 백사면A            이천    경기도  2021    1834.0           10.234   
20     하동본부_3            진주   경상남도  2010     601.0           10.248   
21     삼척소내_2            동해    강원도  2018     990.0           10.255   
22        감우리            충주   충청북도  2021     554.0           10.386   
23      하동변전소            진주   경상남도  2011      48.0           10.732   
24     영월철도부지            영월    강원도  2018    1191.0           11.065   
25     하동본부_5            진주   경상남도  2018     998.0           11.493   
26      하동보건소            진주   경상남도  2011      40.0           11.516   
27     하동본부_4            진주   경상남도  2017     998.0           11.684   
28     하동본부_6            진주   경상남도  2020     998.0           11.824   
29      하동정수장            진주   경상남도  2011      93.0           15.558   

    std_error_rate  sample_count  
0            2.686           602  
1            3.062           252  
2            5.036           380  
3            3.733           148  
4            5.055           142  
5            4.856           686  
6            4.238           148  
7            5.485           264  
8            5.439            98  
9            6.050           374  
10           6.233           262  
11           5.915            98  
12           5.605           142  
13           6.031            96  
14           6.087           264  
15           6.434           264  
16           5.499           182  
17           6.109            94  
18           4.995           144  
19           4.771           142  
20           5.157           144  
21           5.753           264  
22           5.960           150  
23           4.684           148  
24           8.136           260  
25           7.054           148  
26           4.898           146  
27           6.698           148  
28           6.381           148  
29           8.704           146

In [78]:
# =================================================================
# 준공년도 및 용량에 따른 성능 분석
# =================================================================

def analyze_performance_by_numerical_factors(plant_results_df):
    """준공년도와 용량에 따른 성능 분석"""
    
    # 데이터 전처리
    analysis_df = plant_results_df.copy()
    
    # 준공년도를 숫자로 변환 (Unknown 제외)
    analysis_df = analysis_df[analysis_df['준공년도'] != 'Unknown']
    analysis_df['준공년도'] = pd.to_numeric(analysis_df['준공년도'], errors='coerce')
    analysis_df = analysis_df.dropna(subset=['준공년도'])
    
    # 통계 분석
    print("=== 상관관계 분석 ===")
    for data_type in ['valid', 'test', 'external']:
        type_data = analysis_df[analysis_df['data_type'] == data_type].copy()  # .copy() 추가
        
        if len(type_data) > 0:
            corr_year = type_data['준공년도'].corr(type_data['error_rate'])
            corr_capacity = type_data['capacity'].corr(type_data['error_rate'])
            
            print(f"\n{data_type.upper()}:")
            print(f"  준공년도 vs 오차율 상관계수: {corr_year:.3f}")
            print(f"  용량 vs 오차율 상관계수: {corr_capacity:.3f}")
    
    # 시각화
    fig = plt.figure(figsize=(20, 12))
    
    # 1. Scatter plots with regression lines
    for i, data_type in enumerate(['valid', 'test', 'external']):
        type_data = analysis_df[analysis_df['data_type'] == data_type].copy()  # .copy() 추가
        
        if len(type_data) == 0:
            continue
            
        # 준공년도 vs 오차율
        plt.subplot(3, 4, i*4 + 1)
        plt.scatter(type_data['준공년도'], type_data['error_rate'], alpha=0.6, s=30)
        
        # 회귀선 추가
        if len(type_data) > 1:
            z = np.polyfit(type_data['준공년도'], type_data['error_rate'], 1)
            p = np.poly1d(z)
            plt.plot(type_data['준공년도'], p(type_data['준공년도']), "r--", alpha=0.8)
        
        plt.xlabel('Construction Year')
        plt.ylabel('Error Rate (%)')
        plt.title(f'{data_type.upper()} - Year vs Error')
        plt.grid(True, alpha=0.3)
        
        # 용량 vs 오차율 (로그 스케일)
        plt.subplot(3, 4, i*4 + 2)
        plt.scatter(type_data['capacity'], type_data['error_rate'], alpha=0.6, s=30)
        
        # 회귀선 추가
        if len(type_data) > 1:
            z = np.polyfit(np.log(type_data['capacity']), type_data['error_rate'], 1)
            x_smooth = np.logspace(np.log10(type_data['capacity'].min()), 
                                 np.log10(type_data['capacity'].max()), 100)
            y_smooth = z[0] * np.log(x_smooth) + z[1]
            plt.plot(x_smooth, y_smooth, "r--", alpha=0.8)
        
        plt.xlabel('Capacity (MW)')
        plt.ylabel('Error Rate (%)')
        plt.title(f'{data_type.upper()} - Capacity vs Error')
        plt.xscale('log')
        plt.grid(True, alpha=0.3)
        
        # 준공년도 구간별 성능
        plt.subplot(3, 4, i*4 + 3)
        
        # 준공년도를 5년 단위로 그룹화
        type_data['year_group'] = (type_data['준공년도'] // 5) * 5
        year_stats = type_data.groupby('year_group', observed=True)['error_rate'].agg(['mean', 'std', 'count'])
        
        bars = plt.bar(year_stats.index, year_stats['mean'], 
                      yerr=year_stats['std'], capsize=5, alpha=0.7)
        
        # 샘플 수 표시
        for bar, count in zip(bars, year_stats['count']):
            plt.text(bar.get_x() + bar.get_width()/2, bar.get_height() + 0.1,
                    f'n={count}', ha='center', va='bottom', fontsize=8)
        
        plt.xlabel('Construction Year (5-year groups)')
        plt.ylabel('Mean Error Rate (%)')
        plt.title(f'{data_type.upper()} - Error by Year Groups')
        plt.xticks(rotation=45)
        plt.grid(True, alpha=0.3)
        
        # 용량 구간별 성능 (로그 구간)
        plt.subplot(3, 4, i*4 + 4)
        
        # 용량을 로그 구간으로 그룹화
        capacity_bins = [0, 1, 5, 10, 20, 50, 100, 500, float('inf')]
        capacity_labels = ['<1', '1-5', '5-10', '10-20', '20-50', '50-100', '100-500', '500+']
        type_data['capacity_group'] = pd.cut(type_data['capacity'], 
                                           bins=capacity_bins, labels=capacity_labels)
        
        capacity_stats = type_data.groupby('capacity_group', observed=True)['error_rate'].agg(['mean', 'std', 'count'])
        capacity_stats = capacity_stats.dropna()
        
        bars = plt.bar(range(len(capacity_stats)), capacity_stats['mean'], 
                      yerr=capacity_stats['std'], capsize=5, alpha=0.7)
        
        # 샘플 수 표시
        for j, (bar, count) in enumerate(zip(bars, capacity_stats['count'])):
            plt.text(bar.get_x() + bar.get_width()/2, bar.get_height() + 0.1,
                    f'n={count}', ha='center', va='bottom', fontsize=8)
        
        plt.xlabel('Capacity Groups (MW)')
        plt.ylabel('Mean Error Rate (%)')
        plt.title(f'{data_type.upper()} - Error by Capacity Groups')
        plt.xticks(range(len(capacity_stats)), capacity_stats.index, rotation=45)
        plt.grid(True, alpha=0.3)
    
    plt.tight_layout()
    
    # 저장
    save_dir = '/home/local/soominchung_991007/EF/code/tmp/result_analysis/best'
    plt.savefig(os.path.join(save_dir, 'numerical_factors_analysis.png'), dpi=300, bbox_inches='tight')
    plt.close()
    
    # 상세 통계 분석
    detailed_stats = []
    
    for data_type in ['valid', 'test', 'external']:
        type_data = analysis_df[analysis_df['data_type'] == data_type].copy()  # .copy() 추가
        
        if len(type_data) > 0:
            # 준공년도 구간별 통계
            type_data['year_group'] = (type_data['준공년도'] // 5) * 5
            year_stats = type_data.groupby('year_group', observed=True)['error_rate'].agg([
                'count', 'mean', 'std', 'min', 'max'
            ]).round(3)
            year_stats['factor'] = 'construction_year'
            year_stats['data_type'] = data_type
            year_stats['group'] = year_stats.index.astype(str)
            
            # 용량 구간별 통계
            capacity_bins = [0, 1, 5, 10, 20, 50, 100, 500, float('inf')]
            capacity_labels = ['<1', '1-5', '5-10', '10-20', '20-50', '50-100', '100-500', '500+']
            type_data['capacity_group'] = pd.cut(type_data['capacity'], 
                                               bins=capacity_bins, labels=capacity_labels)
            capacity_stats = type_data.groupby('capacity_group', observed=True)['error_rate'].agg([
                'count', 'mean', 'std', 'min', 'max'
            ]).round(3)
            capacity_stats['factor'] = 'capacity'
            capacity_stats['data_type'] = data_type
            capacity_stats['group'] = capacity_stats.index.astype(str)
            
            detailed_stats.extend([year_stats, capacity_stats])
    
    # 통계 결과 DataFrame
    if detailed_stats:
        stats_df = pd.concat(detailed_stats, ignore_index=True)
        stats_df = stats_df.reset_index(drop=True)
        
        print("\n=== 상세 통계 분석 ===")
        print(stats_df.to_string(index=False))
        
        # CSV 저장
        stats_df.to_csv(os.path.join(save_dir, 'numerical_factors_statistics.csv'), index=False)
    
    print(f"\n분석 결과 저장 완료: {save_dir}")
    
    return analysis_df, stats_df if detailed_stats else None
numerical_analysis_df, numerical_stats = analyze_performance_by_numerical_factors(plant_results)    

=== 상관관계 분석 ===

VALID:
  준공년도 vs 오차율 상관계수: 0.193
  용량 vs 오차율 상관계수: 0.097

TEST:
  준공년도 vs 오차율 상관계수: 0.197
  용량 vs 오차율 상관계수: 0.088

EXTERNAL:
  준공년도 vs 오차율 상관계수: -0.045
  용량 vs 오차율 상관계수: 0.389

=== 상세 통계 분석 ===
 count      mean    std   min       max            factor data_type   group
   996  5.050000  3.843 0.267 32.379002 construction_year     valid    2005
  1898  9.896000  5.597 1.386 34.299999 construction_year     valid    2010
  2824  9.658000  6.278 0.102 51.327000 construction_year     valid    2015
   766 10.500000  5.586 0.532 39.166000 construction_year     valid    2020
   294 11.121000  4.799 2.190 32.258999          capacity     valid   20-50
   546  8.780000  7.015 0.727 33.147999          capacity     valid  50-100
  1910  7.185000  4.828 0.267 36.306999          capacity     valid 100-500
  3734 10.001000  6.120 0.102 51.327000          capacity     valid    500+
  1012  4.795000  3.726 0.614 31.857000 construction_year      test    2005
  1954  9.538000  6.200 0.749

In [116]:
# =================================================================
# 준공년도 및 용량에 따른 성능 분석
# =================================================================

def analyze_performance_by_numerical_factors(plant_results_df):
    """준공년도와 용량에 따른 성능 분석"""
    
    # 데이터 전처리
    analysis_df = plant_results_df.copy()
    
    # 준공년도를 숫자로 변환 (Unknown 제외)
    analysis_df = analysis_df[analysis_df['준공년도'] != 'Unknown']
    analysis_df['준공년도'] = pd.to_numeric(analysis_df['준공년도'], errors='coerce')
    analysis_df = analysis_df.dropna(subset=['준공년도'])
    
    # 통계 분석
    print("=== 상관관계 분석 ===")
    for data_type in ['valid', 'test', 'external']:
        type_data = analysis_df[analysis_df['data_type'] == data_type].copy()
        
        if len(type_data) > 0:
            corr_year = type_data['준공년도'].corr(type_data['error_rate'])
            corr_capacity = type_data['capacity'].corr(type_data['error_rate'])
            
            print(f"\n{data_type.upper()}:")
            print(f"  준공년도 vs 오차율 상관계수: {corr_year:.3f}")
            print(f"  용량 vs 오차율 상관계수: {corr_capacity:.3f}")
    
    # 저장 디렉토리
    save_dir = '/home/local/soominchung_991007/EF/code/tmp/result_analysis/best'
    os.makedirs(save_dir, exist_ok=True)
    
    # 1. 준공년도별 성능 분석 (1x3 배열)
    fig, axes = plt.subplots(1, 3, figsize=(18, 6))
    
    for i, data_type in enumerate(['valid', 'test', 'external']):
        type_data = analysis_df[analysis_df['data_type'] == data_type].copy()
        
        if len(type_data) == 0:
            axes[i].set_title(f'{data_type.upper()} - No Data')
            continue
        
        # 준공년도를 5년 단위로 그룹화
        type_data['year_group'] = (type_data['준공년도'] // 5) * 5
        year_stats = type_data.groupby('year_group', observed=True)['error_rate'].agg(['mean', 'std', 'count'])
        
        bars = axes[i].bar(year_stats.index, year_stats['mean'], 
                          yerr=year_stats['std'], capsize=5, alpha=0.7, 
                          color='skyblue', edgecolor='navy', linewidth=1)
        
        # 샘플 수 표시
        for bar, count in zip(bars, year_stats['count']):
            axes[i].text(bar.get_x() + bar.get_width()/2, bar.get_height() + 0.2,
                        f'n={count}', ha='center', va='bottom', fontsize=10, fontweight='bold')
        
        axes[i].set_xlabel('Construction Year (5-year groups)', fontsize=12)
        axes[i].set_ylabel('Mean Error Rate (%)', fontsize=12)
        axes[i].set_title(f'{data_type.upper()} - Error by Construction Year', fontsize=14, fontweight='bold')
        axes[i].tick_params(axis='x', rotation=45, labelsize=11)
        axes[i].tick_params(axis='y', labelsize=11)
        axes[i].grid(True, alpha=0.3)
    
    plt.tight_layout()
    plt.savefig(os.path.join(save_dir, 'construction_year_analysis.png'), dpi=300, bbox_inches='tight')
    plt.close()
    print(f"준공년도별 분석 저장: {save_dir}/construction_year_analysis.png")
    
    # 2. 용량별 성능 분석 (1x3 배열)
    fig, axes = plt.subplots(1, 3, figsize=(18, 6))
    
    for i, data_type in enumerate(['valid', 'test', 'external']):
        type_data = analysis_df[analysis_df['data_type'] == data_type].copy()
        
        if len(type_data) == 0:
            axes[i].set_title(f'{data_type.upper()} - No Data')
            continue
        
        # 용량을 로그 구간으로 그룹화
        capacity_bins = [0, 1, 5, 10, 20, 50, 100, 500, float('inf')]
        capacity_labels = ['<1', '1-5', '5-10', '10-20', '20-50', '50-100', '100-500', '500+']
        type_data['capacity_group'] = pd.cut(type_data['capacity'], 
                                           bins=capacity_bins, labels=capacity_labels)
        
        capacity_stats = type_data.groupby('capacity_group', observed=True)['error_rate'].agg(['mean', 'std', 'count'])
        capacity_stats = capacity_stats.dropna()
        
        bars = axes[i].bar(range(len(capacity_stats)), capacity_stats['mean'], 
                          yerr=capacity_stats['std'], capsize=5, alpha=0.7,
                          color='lightcoral', edgecolor='darkred', linewidth=1)
        
        # 샘플 수 표시
        for j, (bar, count) in enumerate(zip(bars, capacity_stats['count'])):
            axes[i].text(bar.get_x() + bar.get_width()/2, bar.get_height() + 0.2,
                        f'n={count}', ha='center', va='bottom', fontsize=10, fontweight='bold')
        
        axes[i].set_xlabel('Capacity Groups (MW)', fontsize=12)
        axes[i].set_ylabel('Mean Error Rate (%)', fontsize=12)
        axes[i].set_title(f'{data_type.upper()} - Error by Capacity Groups', fontsize=14, fontweight='bold')
        axes[i].set_xticks(range(len(capacity_stats)))
        axes[i].set_xticklabels(capacity_stats.index, rotation=45, ha='right', fontsize=11)
        axes[i].tick_params(axis='y', labelsize=11)
        axes[i].grid(True, alpha=0.3)
    
    plt.tight_layout()
    plt.savefig(os.path.join(save_dir, 'capacity_analysis.png'), dpi=300, bbox_inches='tight')
    plt.close()
    print(f"용량별 분석 저장: {save_dir}/capacity_analysis.png")
    
    # 상세 통계 분석
    detailed_stats = []
    
    for data_type in ['valid', 'test', 'external']:
        type_data = analysis_df[analysis_df['data_type'] == data_type].copy()
        
        if len(type_data) > 0:
            # 준공년도 구간별 통계
            type_data['year_group'] = (type_data['준공년도'] // 5) * 5
            year_stats = type_data.groupby('year_group', observed=True)['error_rate'].agg([
                'count', 'mean', 'std', 'min', 'max'
            ]).round(3)
            year_stats['factor'] = 'construction_year'
            year_stats['data_type'] = data_type
            year_stats['group'] = year_stats.index.astype(str)
            
            # 용량 구간별 통계
            capacity_bins = [0, 1, 5, 10, 20, 50, 100, 500, float('inf')]
            capacity_labels = ['<1', '1-5', '5-10', '10-20', '20-50', '50-100', '100-500', '500+']
            type_data['capacity_group'] = pd.cut(type_data['capacity'], 
                                               bins=capacity_bins, labels=capacity_labels)
            capacity_stats = type_data.groupby('capacity_group', observed=True)['error_rate'].agg([
                'count', 'mean', 'std', 'min', 'max'
            ]).round(3)
            capacity_stats['factor'] = 'capacity'
            capacity_stats['data_type'] = data_type
            capacity_stats['group'] = capacity_stats.index.astype(str)
            
            detailed_stats.extend([year_stats, capacity_stats])
    
    # 통계 결과 DataFrame
    if detailed_stats:
        stats_df = pd.concat(detailed_stats, ignore_index=True)
        stats_df = stats_df.reset_index(drop=True)
        
        print("\n=== 상세 통계 분석 ===")
        print(stats_df.to_string(index=False))
        
        # CSV 저장
        stats_df.to_csv(os.path.join(save_dir, 'numerical_factors_statistics.csv'), index=False)
    
    print(f"\n분석 결과 저장 완료: {save_dir}")
    
    return analysis_df, stats_df if detailed_stats else None

# 실행
numerical_analysis_df, numerical_stats = analyze_performance_by_numerical_factors(plant_results)

=== 상관관계 분석 ===

VALID:
  준공년도 vs 오차율 상관계수: 0.193
  용량 vs 오차율 상관계수: 0.097

TEST:
  준공년도 vs 오차율 상관계수: 0.197
  용량 vs 오차율 상관계수: 0.088

EXTERNAL:
  준공년도 vs 오차율 상관계수: -0.045
  용량 vs 오차율 상관계수: 0.389
준공년도별 분석 저장: /home/local/soominchung_991007/EF/code/tmp/result_analysis/best/construction_year_analysis.png
용량별 분석 저장: /home/local/soominchung_991007/EF/code/tmp/result_analysis/best/capacity_analysis.png

=== 상세 통계 분석 ===
 count      mean    std   min       max            factor data_type   group
   996  5.050000  3.843 0.267 32.379002 construction_year     valid    2005
  1898  9.896000  5.597 1.386 34.299999 construction_year     valid    2010
  2824  9.658000  6.278 0.102 51.327000 construction_year     valid    2015
   766 10.500000  5.586 0.532 39.166000 construction_year     valid    2020
   294 11.121000  4.799 2.190 32.258999          capacity     valid   20-50
   546  8.780000  7.015 0.727 33.147999          capacity     valid  50-100
  1910  7.185000  4.828 0.267 36.306999          cap

해석
준공년도와 오차율의 관계

Valid/Test: 양의 상관관계 (0.193, 0.197)

최근에 지어진 발전소일수록 오차율이 높다
가능한 이유:

최신 발전소는 더 복잡한 기술/시스템을 사용해서 예측이 어려울 수 있음
구형 발전소는 오랜 운영 데이터로 예측 모델이 더 잘 학습되었을 가능성
최신 발전소는 학습 데이터가 상대적으로 적을 수 있음




External: 음의 상관관계 (-0.045)

거의 상관관계 없음 (0에 가까움)
External 데이터는 다른 특성을 가진 것으로 보임



용량과 오차율의 관계

Valid/Test: 약한 양의 상관관계 (0.097, 0.088)

대용량 발전소일수록 약간 오차율이 높다
하지만 상관관계가 매우 약함


External: 중간 정도 양의 상관관계 (0.389)

대용량 발전소의 예측이 훨씬 어렵다
External 데이터에서만 나타나는 특별한 패턴

In [73]:
regional_summary

mean_error  std_error  mean_capacity  total_capacity  plant_count  \
0        9.171      6.351    1319.241943    2.068572e+06            6   
1       10.075      5.187    1364.957031    4.422460e+05            2   
2       10.746      6.033     670.429016    1.174592e+06           12   
3        7.550      5.276     460.227997    1.060366e+06            5   
4        9.595      5.861     680.280029    2.625880e+05            4   
5       10.386      5.960     554.000000    8.310000e+04            1   
6        9.157      7.316    1311.659058    2.025202e+06            6   
7       10.172      6.489    1349.798950    4.562320e+05            2   
8       10.073      6.631     670.466980    1.226954e+06           12   
9        7.431      5.559     484.816010    1.042354e+06            5   
10       9.679      5.064     679.783020    2.691940e+05            4   
11      11.455      9.024     554.000000    7.645200e+04            1   
12      11.129      6.447     482.380005    5.421956e+05            2   
13      14.369     10.256     953.609009    1.378919e+06            3   
14      12.872      7.538     898.132019    1.079554e+06            2   

   data_type   region  
0      valid      강원도  
1      valid      경기도  
2      valid     경상남도  
3      valid       부산  
4      valid    인천광역시  
5      valid     충청북도  
6       test      강원도  
7       test      경기도  
8       test     경상남도  
9       test       부산  
10      test    인천광역시  
11      test     충청북도  
12  external       부산  
13  external    인천광역시  
14  external  제주특별자치도

In [81]:
# =================================================================
# 3. 14시 vs 20시 성능 비교
# =================================================================

def compare_14h_vs_20h_performance():
    """14시와 20시 예측 성능 비교"""
    results = []
    
    for data_type in ['valid', 'test', 'external']:
        data = best_data[data_type]
        
        # 데이터 분할
        total_samples = len(data['pred'])
        split_point = total_samples // 2
        
        # 14시 데이터
        pred_14h = data['pred'][:split_point]
        real_14h = data['real'][:split_point]
        capacity_14h = data['capacity'][:split_point]
        plant_names_14h = data['plant_names'][:split_point]
        
        # 20시 데이터
        pred_20h = data['pred'][split_point:]
        real_20h = data['real'][split_point:]
        capacity_20h = data['capacity'][split_point:]
        plant_names_20h = data['plant_names'][split_point:]
        
        # 각각의 메트릭 계산
        metric_14h = calculate_overall_metrics(pred_14h, real_14h, capacity_14h, plant_names_14h)
        metric_20h = calculate_overall_metrics(pred_20h, real_20h, capacity_20h, plant_names_20h)
        
        # 발전소별 성능 비교
        sample_metrics_14h, valid_samples_14h = get_sample_metrics_with_indices(
            pred_14h, real_14h, capacity_14h, plant_names_14h
        )
        sample_metrics_20h, valid_samples_20h = get_sample_metrics_with_indices(
            pred_20h, real_20h, capacity_20h, plant_names_20h
        )
        
        # 유효한 발전소만 비교 (발전소별로 집계)
        valid_both = valid_samples_14h & valid_samples_20h
        
        # 발전소별로 성능 비교
        plant_comparison = {}
        for i in range(len(plant_names_14h)):
            if valid_both[i]:
                plant_name = plant_names_14h[i]
                if plant_name not in plant_comparison:
                    plant_comparison[plant_name] = {
                        'metric_14h': sample_metrics_14h[i],
                        'metric_20h': sample_metrics_20h[i]
                    }
        
        # 발전소별 카운트
        better_14h_plants = 0
        better_20h_plants = 0
        
        for plant_name, metrics in plant_comparison.items():
            if metrics['metric_14h'] < metrics['metric_20h']:
                better_14h_plants += 1
            elif metrics['metric_20h'] < metrics['metric_14h']:
                better_20h_plants += 1
        
        total_plants_compared = len(plant_comparison)
        
        results.append({
            'data_type': data_type,
            'metric_14h': metric_14h,
            'metric_20h': metric_20h,
            'better_14h_plants': better_14h_plants,
            'better_20h_plants': better_20h_plants,
            'total_plants_compared': total_plants_compared,
            'better_14h_ratio': better_14h_plants / total_plants_compared if total_plants_compared > 0 else 0
        })
    
    comparison_df = pd.DataFrame(results)
    print("\n=== 14시 vs 20시 성능 비교 (발전소 기준) ===")
    print(comparison_df.to_string(index=False))
    
    return comparison_df

comparison_results = compare_14h_vs_20h_performance()


=== 14시 vs 20시 성능 비교 (발전소 기준) ===
data_type  metric_14h  metric_20h  better_14h_plants  better_20h_plants  total_plants_compared  better_14h_ratio
    valid    9.369776    8.868851                 12                 18                     30          0.400000
     test    8.995827    8.952915                 16                 14                     30          0.533333
 external   13.067226   12.785649                  4                  3                      7          0.571429


/tmp/ipykernel_1402427/2295398325.py:25: RuntimeWarning: Mean of empty slice
  sample_metrics = np.nanmean(masked_error, axis=1)
/tmp/ipykernel_1402427/1481092424.py:13: RuntimeWarning: Mean of empty slice
  sample_metrics = np.nanmean(masked_error, axis=1)


In [96]:
# =================================================================
# 4. 계절별 성능 분석 (14h + 20h 통합)
# =================================================================

def load_dates_and_analyze_seasons():
    """계절별 성능 분석 (14h, 20h 통합)"""
    import scipy.stats as stats
    from scipy.stats import f_oneway, kruskal
    
    base_path = '/home/local/soominchung_991007/EF/code/tmp/data/normalized_forecast_modeling_data'
    
    # 날짜 데이터 로드
    dates_14h = {}
    dates_20h = {}
    
    for data_type in ['valid', 'test', 'external_test']:
        # 14시 날짜
        date_path_14h = f"{base_path}/14시/{data_type}/dates.npy"
        try:
            dates_14h[data_type] = np.load(date_path_14h, allow_pickle=True)
        except:
            print(f"Could not load {date_path_14h}")
        
        # 20시 날짜
        date_path_20h = f"{base_path}/20시/{data_type}/dates.npy"
        try:
            dates_20h[data_type] = np.load(date_path_20h, allow_pickle=True)
        except:
            print(f"Could not load {date_path_20h}")
    
    def get_season(date_obj):
        """날짜 객체로부터 계절 반환"""
        try:
            # 배열인 경우 첫 번째 원소 사용 (첫 번째 시간)
            if isinstance(date_obj, np.ndarray):
                date_obj = date_obj[0]
            
            # numpy.datetime64인 경우 pandas Timestamp로 변환
            if isinstance(date_obj, np.datetime64):
                date_obj = pd.Timestamp(date_obj)
            # 문자열인 경우 datetime으로 변환
            elif isinstance(date_obj, str):
                date_obj = datetime.strptime(date_obj, '%Y-%m-%d')
            
            month = date_obj.month
            if month in [3, 4, 5]:
                return 'Spring'
            elif month in [6, 7, 8]:
                return 'Summer'
            elif month in [9, 10, 11]:
                return 'Fall'
            else:
                return 'Winter'
        except Exception as e:
            print(f"Error processing date: {type(date_obj)}, error: {e}")
            return 'Unknown'
    
    # 계절별 분석
    seasonal_results = []
    seasonal_data_for_stats = {}  # 통계 분석용 데이터
    
    for data_type in ['valid', 'test', 'external']:
        # 데이터 타입 매핑 (external -> external_test)
        date_key = 'external_test' if data_type == 'external' else data_type
        
        if date_key not in dates_14h or date_key not in dates_20h:
            print(f"Skipping {data_type}: date data not found")
            continue
            
        data = best_data[data_type]
        total_samples = len(data['pred'])
        split_point = total_samples // 2
        
        print(f"\n=== {data_type.upper()} 데이터 처리 중 ===")
        
        # 14h + 20h 데이터 통합
        combined_pred = np.concatenate([data['pred'][:split_point], data['pred'][split_point:]])
        combined_real = np.concatenate([data['real'][:split_point], data['real'][split_point:]])
        combined_capacity = np.concatenate([data['capacity'][:split_point], data['capacity'][split_point:]])
        combined_plant_names = np.concatenate([data['plant_names'][:split_point], data['plant_names'][split_point:]])
        combined_dates = np.concatenate([dates_14h[date_key], dates_20h[date_key]])
        
        print(f"Combined data shape: {combined_pred.shape}")
        print(f"Combined dates shape: {combined_dates.shape}")
        
        # 각 샘플의 계절 분류
        seasons = []
        for date_sample in combined_dates:
            season = get_season(date_sample)
            seasons.append(season)
        
        # 샘플별 메트릭 계산 (통계 분석용)
        sample_metrics, valid_samples = get_sample_metrics_with_indices(
            combined_pred, combined_real, combined_capacity, combined_plant_names
        )
        
        # 유효한 샘플만 선택
        valid_seasons = np.array(seasons)[valid_samples]
        valid_metrics = sample_metrics[valid_samples]
        
        # 통계 분석용 데이터 저장
        seasonal_data_for_stats[data_type] = {
            'seasons': valid_seasons,
            'metrics': valid_metrics
        }
        
        # 계절별 집계
        for season in ['Spring', 'Summer', 'Fall', 'Winter']:
            season_mask = valid_seasons == season
            season_count = np.sum(season_mask)
            
            if season_count == 0:
                continue
            
            season_metrics = valid_metrics[season_mask]
            season_mean = np.mean(season_metrics)
            
            print(f"  {season}: {season_count} samples, mean error: {season_mean:.3f}%")
            
            seasonal_results.append({
                'data_type': data_type,
                'season': season,
                'metric': season_mean,
                'sample_count': season_count,
                'std': np.std(season_metrics)
            })
    
    seasonal_df = pd.DataFrame(seasonal_results)
    print("\n=== 계절별 성능 분석 ===")
    print(seasonal_df.to_string(index=False))
    
    # 통계 분석
    print("\n=== 계절별 성능 통계 분석 ===")
    
    for data_type in ['valid', 'test', 'external']:
        if data_type not in seasonal_data_for_stats:
            continue
            
        seasons = seasonal_data_for_stats[data_type]['seasons']
        metrics = seasonal_data_for_stats[data_type]['metrics']
        
        # 계절별 데이터 분리
        spring_data = metrics[seasons == 'Spring']
        summer_data = metrics[seasons == 'Summer']
        fall_data = metrics[seasons == 'Fall']
        winter_data = metrics[seasons == 'Winter']
        
        season_groups = [spring_data, summer_data, fall_data, winter_data]
        season_names = ['Spring', 'Summer', 'Fall', 'Winter']
        
        # 비어있지 않은 그룹만 선택
        valid_groups = []
        valid_names = []
        for group, name in zip(season_groups, season_names):
            if len(group) > 0:
                valid_groups.append(group)
                valid_names.append(name)
        
        if len(valid_groups) >= 2:
            # ANOVA 검정 (정규성 가정)
            try:
                f_stat, p_value_anova = f_oneway(*valid_groups)
                print(f"\n{data_type.upper()} - ANOVA 검정:")
                print(f"  F-statistic: {f_stat:.4f}")
                print(f"  p-value: {p_value_anova:.4f}")
                print(f"  결과: {'계절별 차이 있음' if p_value_anova < 0.05 else '계절별 차이 없음'} (α=0.05)")
            except:
                print(f"\n{data_type.upper()} - ANOVA 검정 실패")
            
            # Kruskal-Wallis 검정 (비모수)
            try:
                kw_stat, p_value_kw = kruskal(*valid_groups)
                print(f"\n{data_type.upper()} - Kruskal-Wallis 검정:")
                print(f"  H-statistic: {kw_stat:.4f}")
                print(f"  p-value: {p_value_kw:.4f}")
                print(f"  결과: {'계절별 차이 있음' if p_value_kw < 0.05 else '계절별 차이 없음'} (α=0.05)")
            except:
                print(f"\n{data_type.upper()} - Kruskal-Wallis 검정 실패")
            
            # 기술통계
            print(f"\n{data_type.upper()} - 계절별 기술통계:")
            for group, name in zip(valid_groups, valid_names):
                print(f"  {name}: n={len(group)}, mean={np.mean(group):.3f}, std={np.std(group):.3f}")
    
    # 시각화 (계절별 색상)
    season_colors = {
        'Spring': '#90EE90',  # 연두색
        'Summer': '#FFD700',  # 노란색
        'Fall': '#FF8C00',    # 주황색
        'Winter': '#87CEEB'   # 하늘색
    }
    
    fig, axes = plt.subplots(1, 3, figsize=(18, 6))
    
    for i, data_type in enumerate(['valid', 'test', 'external']):
        subset = seasonal_df[seasonal_df['data_type'] == data_type]
        
        if len(subset) > 0:
            # 계절 순서 고정
            seasons_order = ['Spring', 'Summer', 'Fall', 'Winter']
            subset_ordered = []
            colors_ordered = []
            
            for season in seasons_order:
                season_data = subset[subset['season'] == season]
                if len(season_data) > 0:
                    subset_ordered.append(season_data.iloc[0])
                    colors_ordered.append(season_colors[season])
            
            if subset_ordered:
                subset_ordered = pd.DataFrame(subset_ordered)
                
                bars = axes[i].bar(subset_ordered['season'], subset_ordered['metric'], 
                                 color=colors_ordered, alpha=0.8, edgecolor='black', linewidth=1)
                
                # 에러바 추가
                axes[i].errorbar(subset_ordered['season'], subset_ordered['metric'], 
                               yerr=subset_ordered['std'], fmt='none', color='black', capsize=5)
                
                # 샘플 수 표시
                for bar, count in zip(bars, subset_ordered['sample_count']):
                    axes[i].text(bar.get_x() + bar.get_width()/2, bar.get_height() + 0.1,
                                f'n={count}', ha='center', va='bottom', fontsize=10, fontweight='bold')
                
                axes[i].set_title(f'{data_type.upper()} - Seasonal Performance', fontsize=14, fontweight='bold')
                axes[i].set_ylabel('Error Rate (%)', fontsize=12)
                axes[i].set_xlabel('Season', fontsize=12)
                axes[i].grid(True, alpha=0.3)
                axes[i].tick_params(axis='both', labelsize=11)
        else:
            axes[i].set_title(f'{data_type.upper()} - No Data', fontsize=14)
    
    plt.tight_layout()
    
    # 저장
    save_dir = '/home/local/soominchung_991007/EF/code/tmp/result_analysis/best'
    plt.savefig(os.path.join(save_dir, 'seasonal_analysis_combined.png'), dpi=300, bbox_inches='tight')
    plt.close()
    
    # CSV 저장
    seasonal_df.to_csv(os.path.join(save_dir, 'seasonal_results_combined.csv'), index=False)
    
    print(f"\n결과 저장 완료: {save_dir}")
    
    return seasonal_df

# 계절별 분석 실행
seasonal_results = load_dates_and_analyze_seasons()


=== VALID 데이터 처리 중 ===
Combined data shape: (6812, 24)
Combined dates shape: (6812, 24)
  Spring: 1140 samples, mean error: 8.132%
  Summer: 2276 samples, mean error: 9.405%
  Fall: 2238 samples, mean error: 9.858%
  Winter: 830 samples, mean error: 7.701%

=== TEST 데이터 처리 중 ===
Combined data shape: (6846, 24)
Combined dates shape: (6846, 24)
  Spring: 1224 samples, mean error: 7.462%
  Summer: 914 samples, mean error: 7.792%
  Fall: 2492 samples, mean error: 9.428%
  Winter: 1766 samples, mean error: 9.993%

=== EXTERNAL 데이터 처리 중 ===
Combined data shape: (4034, 24)
Combined dates shape: (4034, 24)
  Spring: 1232 samples, mean error: 12.683%
  Summer: 734 samples, mean error: 11.481%
  Fall: 746 samples, mean error: 12.335%
  Winter: 1060 samples, mean error: 14.627%

=== 계절별 성능 분석 ===
data_type season    metric  sample_count       std
    valid Spring  8.131905          1140  5.754978
    valid Summer  9.404975          2276  5.657555
    valid   Fall  9.857811          2238  6.22070

/tmp/ipykernel_1402427/1481092424.py:13: RuntimeWarning: Mean of empty slice
  sample_metrics = np.nanmean(masked_error, axis=1)



결과 저장 완료: /home/local/soominchung_991007/EF/code/tmp/result_analysis/best


계절별 태양광 발전 예측 성능 분석 결과
통계적 유의성

모든 데이터셋에서 계절별 성능 차이가 통계적으로 매우 유의함 (p < 0.001)
ANOVA와 Kruskal-Wallis 검정 모두 일관된 결과

계절별 성능 패턴
VALID 데이터

겨울 > 봄 > 여름 > 가을 순으로 성능 좋음
겨울(7.7%) vs 가을(9.9%): 약 2.2%p 차이

TEST 데이터

봄 > 여름 > 가을 > 겨울 순으로 성능 좋음
봄(7.5%) vs 겨울(10.0%): 약 2.5%p 차이

EXTERNAL 데이터

여름 > 가을 > 봄 > 겨울 순으로 성능 좋음
여름(11.5%) vs 겨울(14.6%): 약 3.1%p 차이

핵심 인사이트
겨울철 예측이 가장 어려운 경우가 많음 (TEST, EXTERNAL에서 최악 성능)
봄/여름이 상대적으로 예측하기 쉬운 계절인 경향

### loss 잘 돌아갔나 Sample로 확인

In [22]:
# 첫 번째 폴더의 데이터로 샘플 테스트
sample_folder = list(valid_files.keys())[0]
sample_data = valid_files[sample_folder]

print(f"=== {sample_folder} 샘플 데이터 메트릭 계산 과정 ===")
print(f"pred shape: {sample_data['pred'].shape}")
print(f"real shape: {sample_data['real'].shape}")
print(f"capacity shape: {sample_data['capacity'].shape}")

# 첫 번째 샘플만 선택해서 과정 확인
pred_sample = sample_data['pred'][0]  # 첫 번째 샘플
real_sample = sample_data['real'][0]
capacity_sample = sample_data['capacity'][0]

print(f"\n--- 첫 번째 샘플 ---")
print(f"pred_sample shape: {pred_sample.shape}")
print(f"real_sample shape: {real_sample.shape}")
print(f"capacity_sample: {capacity_sample}")

print('real sample')
print(real_sample,'\n')
# 이용률 계산
utilization_rate = (real_sample / capacity_sample) * 100
print(f"\n이용률: {utilization_rate}")
print(f"10% 이상 유효한 시점 수: {np.sum(utilization_rate >= 10.0)}/{len(utilization_rate)}")

# 오차율 계산
error_rate = np.abs(real_sample - pred_sample) / capacity_sample * 100
print(f"\n오차율: {error_rate}")

# 유효한 시점만 선택
valid_mask = utilization_rate >= 10.0
masked_error = np.where(valid_mask, error_rate, np.nan)
print(f"\n유효한 시점의 오차율: {masked_error[~np.isnan(masked_error)]}")

# 샘플 메트릭 (시간축 평균)
sample_metric = np.nanmean(masked_error)
print(f"\n이 샘플의 평균 오차율: {sample_metric}")

# 전체 샘플들 계산
capacity_expanded = sample_data['capacity'][:, np.newaxis]
utilization_rate_all = (sample_data['real'] / capacity_expanded) * 100
valid_mask_all = utilization_rate_all >= 10.0
error_rate_all = np.abs(sample_data['real'] - sample_data['pred']) / capacity_expanded * 100
masked_error_all = np.where(valid_mask_all, error_rate_all, np.nan)
sample_metrics_all = np.nanmean(masked_error_all, axis=1)

print(f"\n=== 전체 샘플 결과 ===")
print(f"전체 샘플 수: {len(sample_metrics_all)}")
print(f"유효한 샘플 수: {np.sum(~np.isnan(sample_metrics_all))}")
print(f"전체 평균 오차율: {np.nanmean(sample_metrics_all)}")

=== simple_128d_model_1n_layers_0.0001lr_32batch_SmoothL1Loss_0712_140324 샘플 데이터 메트릭 계산 과정 ===
pred shape: (6812, 24)
real shape: (6812, 24)
capacity shape: (6812,)

--- 첫 번째 샘플 ---
pred_sample shape: (24,)
real_sample shape: (24,)
capacity_sample: 603.0
real sample
[  0.      0.      0.      0.      0.      0.      0.      0.482  31.337
  94.975 153.792 335.545 442.573 464.267 452.215 416.057 339.884 232.375
  81.958   7.714   0.      0.      0.      0.   ] 


이용률: [ 0.          0.          0.          0.          0.          0.
  0.          0.07993367  5.196849   15.750414   25.504478   55.645935
 73.395195   76.99287    74.994194   68.99785    56.36551    38.536484
 13.591708    1.2792704   0.          0.          0.          0.        ]
10% 이상 유효한 시점 수: 10/24

오차율: [ 0.          0.          0.          0.          0.          0.
  0.          0.07993367  0.54315627  4.411134   14.466569    0.8749384
 16.335928   18.8551     19.516756   23.234419   28.640604   25.931734
 13.591708 

/tmp/ipykernel_1402427/1743386814.py:46: RuntimeWarning: Mean of empty slice
  sample_metrics_all = np.nanmean(masked_error_all, axis=1)
